daemon: add --no-volumes flag

runcom opened the pull request:
Prevents running containers with volumes. This patch will block running
containers from images built with VOLUME(s) instructions. It will also
block running cotainers that define a volume (docker run -v). docker
create -v won't error out instead.
This patch adds a --no-volumes flag to the docker daemon which prevents
what said above. The feature is opt-in, otherwise the engine will run
as it was.
This will allow operators to allow users to use their own images but
want to control where users can write.

Signed-off-by: Antonio Murdaca runcom@redhat.com

calavera commented:
I'm worried that this proposal sounds like a workaround for something else.

If I understand this change correctly, it allows people to explicitly run images that they know require volumes without actually mounting those volumes. I'd really wonder why that image was built with a volume requirement in the first place if you actively don't want to mount that volume.

I'm leaning towards not adding this flag, but I'd like to hear other opinions.

runcom commented:
@calavera this will fully block a docker run from an image which has a volume built in and this is to allow operators to restrict user to write to persistent storage (there's no control right now, one can potentially fill up a whole disk by just running a container)

thaJeztah commented:
This also makes `docker build` act completely different on systems where it's disabled than on systems where it isn't.

w.r.t restricting storage / usage of volumes, perhaps that can be regulated once the authz PR comes through and the builder is moved to the client (i.e. the authz plugin can block certain actions / API calls). For non bind-mounted volumes, we can consider a `--default-volume-driver` at some point, allowing a custom volume plugin with quota support. (note there's an open issue for quota support in general; https://github.com/docker/docker/issues/3804)

I do still think we should have a way to reset properties from a parent image though (e.g. the `UNSET` / `UNVOLUME` proposal that we had).

duglin commented:
I can see a need to restrict this a `docker run` time.

I'm still a bit fuzzy on the `docker build` time though. I don't think the warnings are appropriate because warning them about what MIGHT happen at runtime seems pointless since we have no idea whether this flag will be turned on or not when this image is run.  So, telling the builder (instead of the person) running the image does no good - they'll learn to just ignore the warning. But, erroring out at runtime when this flag it on and the image has volumes does make sense.

runcom commented:
@thaJeztah quota support could be the end goal but right now there's no way to restrict this. Not sure about the authz, but probably could be helpful in blocking certain request (even if there will be logic related to the image being run) 

@duglin you're right the warning is practically useless.. I just wanted to make sure who builds images get warned about the possibility to not being able to run the image - when she built it. I'll remove the warning though

thaJeztah commented:
For runtime, perhaps `--volume-driver=none`?

However, I think the biggest issue is the presence of a `VOLUME` declaration in base images, which make them very impractical to build upon (use case; building e.g. MySQL image with a default database propagated).

runcom commented:
@thaJeztah I'm updating the patch but all of this applies also to `docker run -v` which creates a volume. This is to say it's not so tied to the `VOLUME` instruction and we want to block any run which uses volumes

runcom committed:
daemon: add --no-volumes flag

Prevents running containers with volumes. This patch will block running
containers from images built with VOLUME(s) instructions. It will also
block running cotainers that define a volume (docker run -v). docker
create -v won't error out instead.
This patch adds a --no-volumes flag to the docker daemon which prevents
what said above. The feature is opt-in, otherwise the engine will run
as it was.
This will allow operators to allow users to use their own images but
want to control where users can write.

Signed-off-by: Antonio Murdaca <runcom@redhat.com>

runcom commented:
@thaJeztah I've updated the patch to reflect my use case

cpuguy83 commented:
Definitely -1 on this.
This can be done using the new authorization framework instead.

runcom commented:
> Definitely -1 on this.
> This can be done using the new authorization framework instead.

Working on that..

runcom commented:
I've created https://github.com/runcom/dkauthz (credits to @calavera) and implemented the same with a plugin

calavera commented:
can we close this PR then @runcom ?

runcom commented:
@calavera closed

thaJeztah commented:
> I've created https://github.com/runcom/dkauthz (credits to @calavera) and implemented the same with a plugin

Awesome! You can create a PR to add it to https://github.com/docker/docker/blob/master/docs/extend/plugins.md :innocent:
